### PR TITLE
Implement session state switch infrastructure required for DevExpress/testcafe#1454

### PR DESCRIPTION
**Pull request by LavrovArtem**
/cc @miherlosev @AlexanderMoskovkin

### CODE REVIEW

File: src/processing/resources/page.js [hunk 1]
```diff
@@ -33,6 +34,22 @@ class PageProcessor extends ResourceProcessorBase {
         this.serializer = new parse5.Serializer();
     }
 
+    _createRestoreStoragesScript (storageKey, storages) {
+        const scriptStr = dedent(`
+            <script type="text/javascript" class="${ SHADOW_UI_CLASSNAME.selfRemovingScript }">
```
Comment (miherlosev) [inline src/processing/resources/page.js]: There are two self removing scripts ( `BODY_CREATED_EVENT_SCRIPT` and restore storages script) with shared client code: tag, self-executing function and remove scripts parts.
Let's create the utility function `createSelfRemovingScript (content)` and use it for both scripts.

File: src/session/index.js [hunk 2]
```diff
@@ -121,9 +130,10 @@ export default class Session extends EventEmitter {
         this.externalProxySettings = settings;
     }
 
-    onPageRequest () {
+    onPageRequest (ctx) {
         if (this.requireStateSwitch) {
-            this.cookies.setJar(this.pendingStateSnapshot);
+            this.cookies.setJar(this.pendingStateSnapshot.cookies);
+            ctx.restoredStorages      = this.pendingStateSnapshot.storages;
```
Comment (AlexanderMoskovkin) [inline src/session/index.js]: I think `restoringStorage` is more appropriate

Comment (LavrovArtem) [inline src/session/index.js]: Do you mean `restoringStorages`?

File: src/client/sandbox/storages/index.js [hunk 3]
```diff
@@ -35,7 +30,7 @@ export default class StorageSandbox extends SandboxBase {
     _createStorageWrappers () {
         const sessionId             = settings.get().sessionId;
         const host                  = destLocation.getParsed().host;
-        const storageKey            = StorageSandbox._getStorageKey(sessionId, host);
+        const storageKey            = getStorageKey(sessionId, host);
```
Comment (miherlosev) [inline src/client/sandbox/storages/index.js]: For `StorageSandbox.backup` method we should use own `JSON` implementation

File: src/processing/resources/page.js [hunk 4]
```diff
@@ -33,6 +34,22 @@ class PageProcessor extends ResourceProcessorBase {
...
+                (function () {
+                    window.localStorage.setItem("${ storageKey }", ${ JSON.stringify(storages.localStorage) });
```
Comment (miherlosev) [inline src/processing/resources/page.js]: This script can have a large size. Need to run health monitor on this commit

File: src/client/sandbox/storages/index.js (hunk 3 above)
Comment (LavrovArtem) [inline src/client/sandbox/storages/index.js]: done

File: src/processing/resources/page.js (hunk 1 above)
Comment (LavrovArtem) [inline src/processing/resources/page.js]: We have a four self removing scripts. I think it is task for a separate issue.

Comment (miherlosev) [inline src/processing/resources/page.js]: Ok. create an issue for this

Comment (LavrovArtem) [inline src/processing/resources/page.js]: https://github.com/DevExpress/testcafe-hammerhead/issues/1229

File: src/session/index.js (hunk 2 above)
Comment (AlexanderMoskovkin) [inline src/session/index.js]: Yep

**Comment by LavrovArtem** [comment]
FPR

**Pull request merged**